### ISSUE TITLE
Proposal for `$return` extension

### ISSUE THREAD

**Issue by Cheatoid** [TypeScriptToLua/TypeScriptToLua#1120]
### Problem description
TypeScript does not allow `return` outside of function body (in my opinion, this is a severe limitation).
```typescript
if (os.time() % 2 == 0) {
    return $multi("foo", math.pi); // 'return' statement can only be used within a function body
}
```

### Suggested solution
Therefore I propose to introduce a new extension (`$return`) to remedy this limitation easily:
```typescript
if (os.time() % 2 == 0) {
    $return("foo", math.pi);
}
```

#### Expected Lua translation
This is the goal (in this case), when the above snippet is transpiled, it should produce exactly this output:
```lua
if os.time() % 2 == 0 then
    return "foo", math.pi
end
```
No more, no less.

#### Details
`$return` should work just like [`$multi`](https://typescripttolua.github.io/docs/advanced/language-extensions#multi), accepting multiple arguments (since Lua can return many values), with one important difference being that it would emit the `return` keyword/statement, of course.

In other words, `$return` would allow TSTL to return plain values directly from Lua script on *main* scope, which is otherwise not possible!

**Comment by hazzard993** [comment]
There is an `export =` syntax in TypeScript to allow modules to return plain values

```ts
export = [1, 2, 3];
// this modules now returns [1, 2, 3], nothing else
```

You may need to enable the `esModuleInterop` flag when importing a module that exports using this syntax since the import syntax could work in both ways

```ts
import values from "./array";
import * as values from "./array";
```

Lua also only stores the first returned value from a module so `return "foo", math.pi` means that only `"foo"` is returned from the module

If you want if statements to effect the return value you can do something like this too

```ts
let out: any;

if (cond) {
  out = { ... };
}

export = out;
```

**Comment by Cheatoid** [comment]
> There is an `export =` syntax in TypeScript to allow modules to return plain values

Already tried that and this doesn't work, as described in my post (once again due to TypeScript limitation):
```typescript
if (os.time() % 2 == 0) {
    export = ["foo", math.pi];
}
```
Results in: "An export assignment must be at the top level of a file or module declaration."

> You may need to enable the `esModuleInterop` flag when importing a module that exports using this syntax since the import syntax could work in both ways

This is actually irrelevant in my practical usecase, since it is a (game) engine loading and running a Lua script, and all I care about is that the transpiled Lua code must return plain values directly on *main* scope under certain conditions (when I say plain, I mean it should not have values wrapped in a table).

> If you want if statements to effect the return value you can do something like this too [...]

Would that not wrap the plain values into a table? I think you don't fully understand my proposal.
I just double-checked this, and yeah that would wrap it in a table, which is **unwanted**, let me remind you I have explicitly said "No more, no less".
Short said, none what you said works; It does not meet the expected Lua translation that I have provided.

Please read proposal and try again.

**Comment by hazzard993** [comment]
> Already tried that and this doesn't work, as described in my post (once again due to TypeScript limitation):

No, you can't use `export =` inside a block. So see my **second** example

```ts
let out: any;

if (cond) {
  out = ["foo", math.pi];
}

export = out;
```

Yes, `out` is a table. But it was to show you how to use `export =` and an if statement to export a `plain value`.

> This is actually irrelevant in my practical usecase, since it is a (game) engine loading and running a Lua script, and all I care about is that the transpiled Lua code must return plain values directly on main scope under certain conditions

An unmentioned use case?

To re-iterate, `Lua also only stores the first returned value from a module`.so give us an idea of this Lua environment you're using, how big of an issue is this?

> Short said, none of what you said works; It does not meet the expected Lua translation that I have provided.
>
> Please read proposal and try again.

You tried out `export =` once, failed to use it correctly, and then blamed the code.

Here's some more ideas! Read them

```ts
// @ts-ignore
return $multi
```

Or this?

```ts
export = (() => {
  if (cond) {
    return $multi("foo", math.pi);
  }
})();
```

Currently both these examples

**Comment by YoRyan** [comment]
For the record, the `export =` syntax does not wrap the returned value in a table:
```typescript
export = 42;
```
```lua
local ____exports = 42
return ____exports
```
But TypeScript only allows that syntax in commonjs mode. If it's in es module mode, it'll only permit the use of `export default`, which indeed creates a table:
```lua
local ____exports = {}
____exports.default = 42
return ____exports
```
As @hazzard993 said, Lua might allow a file-level return with multiple values, but it discards all but the first. (Try it.)

So with respect to this proposal, the only missing thing in TSTL is the ability to make a file-level return from anywhere else in the code.
